src/index.ts

```diff
@@ -327,52 +255,16 @@ export class Paginator {
    *     and returns `nextQuery` to receive more results.
    * @return {stream} - Readable object stream.
    */
-  runAsStream_(parsedArguments: ParsedArguments, originalMethod: Function) {
-    const query = parsedArguments.query;
-    let resultsToSend = parsedArguments.maxResults!;
-
-    const limiter = exports.createLimiter(makeRequest, {
-      maxApiCalls: parsedArguments.maxApiCalls,
-      streamOptions: parsedArguments.streamOptions,
-    });
-
-    const stream = limiter.stream;
-
-    stream.once('reading', () => {
-      limiter.makeRequest(query);
-    });
-
-    function makeRequest(query?: ParsedArguments | string) {
-      originalMethod(query, onResultSet);
-    }
-
-    // tslint:disable-next-line:no-any
-    function onResultSet(err: Error | null, results?: any[], nextQuery?: any) {
-      if (err) {
-        stream.destroy(err);
-        return;
-      }
-
-      if (resultsToSend >= 0 && results!.length > resultsToSend) {
-        results = results!.splice(0, resultsToSend);
-      }
-
-      resultsToSend -= results!.length;
-
-      split(results!, stream).then(streamEnded => {
-        if (streamEnded) {
-          return;
-        }
-        if (nextQuery && resultsToSend !== 0) {
-          limiter.makeRequest(nextQuery);
-          return;
-        }
-        stream.push(null);
-      });
-    }
-    return limiter.stream;
+  // tslint:disable-next-line:no-any
+  runAsStream_<T = any>(
+    parsedArguments: ParsedArguments,
+    originalMethod: Function
+  ): ResourceStream<T> {
+    return new ResourceStream<T>(parsedArguments, originalMethod);
```

Comment:
This is a rough draft attempt at an option for how this could look if split-array-stream changes go through:

```ts
runAsStream_ (parsedArguments, originalMethod) {
  let numMaxApiCalls = args.maxApiCalls === -1 ? Infinity : args.maxApiCalls;
  let numRequestsMade;
  let numResultsToSend = args.maxResults === -1 ? Infinity  : args.maxResults;

  let query = args.query;

  const splitStream = new SplitArrayStream(getPage);

  const getPage = () => {
    return new Promise((resolve, reject) => {
      originalMethod(query, (err, results, nextQuery) => {
        if (err) {
          splitStream.destroy(err);
          return;
        }

        if (numResultsToSend !== Infinity) {
          results = results.splice(0, numResultsToSend);
          _numResultsToSend -= results.length;
        }

        const didMakeMaxCalls = ++numRequestsMade >= numMaxApiCalls;

        if (didMakeMaxCalls) {
          results.push(null);
        } else {
          query = nextQuery;
        }

        resolve(results);
      });
    });
  };

  return splitStream;
}
```

Reply:
@stephenplusplus do you have a timeline of when this would be finished?

Reply:
I can probably wrap up https://github.com/stephenplusplus/split-array-stream/pull/4 tomorrow.